enhancements/machine-config/admin-defined-reboot-policy.md

```diff
@@ -0,0 +1,264 @@
+
+The MachineConfigNodes CR will track the status of the upgrade, and will also be updated to indicate what the RebootPolicy used for the update was.
+
+RebootPolicies are only processed in-cluster and cannot be used to skip the bootstrap pivot during cluster installation.
+
+#### Variation and form factor considerations [optional]
+
+This will work by default on standalone OCP and any other OCP variants that run the MCO in its entirety. (e.g. works for RHEL)
+
+HyperShift deployments do not run the MCO pods directly and will not have RebootPolicy objects.
+
+### API Extensions
+
+- Adds the RebootPolicy CRD (only under MCO namespace and only affects MCO)
+
+### Implementation Details/Notes/Constraints [optional]
+
+This aims more to be a user-facing API to essentially map MachineConfig fields to what action should be taken on them, which aligns with existing MCO functionality to perform a hard-coded list of these policies. However, as we evolve the MCO to fit in with an image based OS update vision, we would need this to work with:
+ - Existing MachineConfig based entries
+ - User-provided base-image upgrade cases (live-apply)
+ - A more unified update interface with bootc
+
+Which we can definitely translate under-the-hood to adapt to the system, but file/unit specific entries like this may not work as effectively when it comes to image-to-image updates, especially at the daemon level.
+
+As there is a need for this in 4.16, we will proceed with the format that works best with the existing MCO architecture, while leaving it open to change for the future.
+
+#### Hypershift [optional]
+
+As noted above, Hypershift cannot directly tap into this. In a bit more detail, Hypershift has 2 modes of operation for nodepools:
+Replace upgrades - since all machines are deleted and recreated in this model, there is no method of live-application today
+Inplace upgrades - these actually use a special form of the daemon, and we can leverage that to update and perform necessary RebootPolicies. However we probably do not want hosted cluster admins to have direct control over this, much like MachineConfigs, so if we do want to add this we will need a new API in Hypershift to handle this. As such, we are not considering this at this time
+
+### Risks and Mitigations
+
+As noted in the non-goal sections, any RebootPolicy is not verified in the MCO, minus formatting. This means that if you accidentally define a change that needed a reboot to take effect, the cluster would “finish the update” without being able to use the new configuration.
+
+The user thereby needs to understand that no additional verification is taking place, and that they accept any responsibility for RebootPolicies causing configuration not to take effect.
```

Comment:
How will you explain this to a user?

Reply:
Docs, hopefully :)

We can craft an alert or something to indicate when their rebootpolicy took effect upon an update. Without constraining the user to set based on a list of tested "safe" policies (Which goes back to what we already have today, and we don't want to continue extending necessarily) there isn't a way to guarantee (that I know of) user-defined policies to be always "Safe"

Reply:
Providing some metric or alert that is easy for support to spot I think would have value